### PR TITLE
Add fuzzing application for gcoap

### PR DESCRIPTION
**Pull request by nmeum**
### Contribution description

Based on #13157, this PR adds a fuzzing application for the `gcoap` module. Since it's initial implementation in 2019 the `gcoap` module switched to using `sock_async_event`. For this reason, this PR also includes an implementation of a fuzzing termination condition for `sock_async_event`. After the callback has been invoked, the code checks if the previous packet returned by `gnrc_sock_recv` was the fuzzing packet, if so the fuzzing application is terminated.

### Testing procedure

One function tested extensively by this fuzzing application is `coap_parse()`. As such, this fuzzing application would have been capable of finding #10753. To test this, remove the following code from the `coap_parse()` function:

https://github.com/RIOT-OS/RIOT/blob/b6be8af81c3f7f38b00340d6441d08a34431d865/sys/net/application_layer/nanocoap/nanocoap.c#L111-L114

Afterwards run:

    $ make -C fuzzing/gcoap all-asan
    $ make -C fuzzing/gcoap fuzz

AFL should discover a crash, in the modified code branch, within a few minutes.

### Issues/PRs references

* #13157 
* #10753

### CODE REVIEW

File: sys/net/sock/async/event/sock_async_event.c [hunk 1]
```diff
@@ -36,6 +41,12 @@ static inline void _cb(void *sock, sock_async_flags_t type, void *arg,
     ctx->event.cb_arg = arg;
     ctx->event.type |= type;
     event_post(ctx->queue, &ctx->event.super);
+
+#ifdef MODULE_FUZZING
```
Comment (benpicco) [inline sys/net/sock/async/event/sock_async_event.c:45]: Please add a comment.

File: sys/net/gnrc/sock/gnrc_sock.c [hunk 2]
```diff
@@ -90,8 +91,7 @@ ssize_t gnrc_sock_recv(gnrc_sock_reg_t *reg, gnrc_pktsnip_t **pkt_out,
     msg_t msg;
 
 #ifdef MODULE_FUZZING
-    static gnrc_pktsnip_t *prevpkt;
-    if (prevpkt && prevpkt == gnrc_pktbuf_fuzzptr) {
+    if (gnrc_sock_prevpkt && gnrc_sock_prevpkt == gnrc_pktbuf_fuzzptr) {
```
Comment (benpicco) [inline sys/net/gnrc/sock/gnrc_sock.c:94]: Please add a comment here too - missed that in the last review.

File: sys/net/gnrc/sock/gnrc_sock.c [hunk 3]
```diff
@@ -90,8 +91,7 @@ ssize_t gnrc_sock_recv(gnrc_sock_reg_t *reg, gnrc_pktsnip_t **pkt_out,
     msg_t msg;
 
 #ifdef MODULE_FUZZING
```
Comment (nmeum) [inline sys/net/gnrc/sock/gnrc_sock.c:93]: Should probably also be changed to `if defined(MODULE_FUZZING) && !defined(MODULE_SOCK_ASYNC_EVENT)`.

**Comment by benpicco** [comment]
This is unrelated but I've noticed `make -j fuzz` will only ever use one core.
Is this a limitation of `afl`? To me it seems such task would lend itself well to parallelization.

**Comment by nmeum** [comment]
> This is unrelated but I've noticed `make -j fuzz` will only ever use one core.

AFL has it's own parallel fuzzing setup. See:

* https://github.com/RIOT-OS/RIOT/tree/master/fuzzing#parallel-fuzzing
* https://github.com/google/AFL/blob/master/docs/parallel_fuzzing.txt

File: sys/net/gnrc/sock/gnrc_sock.c (hunk 3 above)
Comment (benpicco) [reply sys/net/gnrc/sock/gnrc_sock.c:93]: Or even better

```C
if (IS_ACTIVE(MODULE_FUZZING) && !IS_ACTIVE(MODULE_SOCK_ASYNC_EVENT)) {
    …
}
```

**Comment by benpicco** [comment]
@nmeum ping :wink:

**Comment by nmeum** [comment]
Having thought about this further commit 7d1cf1dc3bdba11bfcceca44df49e54df1633a06 is probably not needed since `MODULE_SOCK_ASYNC_EVENT` will also cause `MODULE_GNRC_SOCK` to be defined for fuzzing applications.

**Comment by benpicco** [comment]
Well feel free to drop it then :wink:

**Pull request merged**